src/main/scala/com/codacy/analysis/cli/upload/ResultsUploader.scala

```diff
@@ -1,13 +1,40 @@
 package com.codacy.analysis.cli.upload
 
+import cats.implicits._
 import com.codacy.analysis.cli.clients.CodacyClient
 import com.codacy.analysis.cli.model.Result
 import org.log4s.{Logger, getLogger}
 
 import scala.concurrent.{ExecutionContext, Future}
 import scala.util.{Failure, Success}
 
-class ResultsUploader(commitUuid: String, codacyClient: CodacyClient, batchSizeOpt: Option[Int])(
+object ResultsUploader {
+
+  private val logger: Logger = getLogger
+
+  def apply(codacyClientOpt: Option[CodacyClient],
+            upload: Boolean,
+            commitUuid: Option[String],
+            batchSize: Option[Int] = Option.empty[Int])(
+    implicit context: ExecutionContext): Either[String, Option[ResultsUploader]] = {
+    if (upload) {
+      codacyClientOpt.fold {
```

Comment:
I'd favour a for comprehension here, instead of a nested fold.

Reply:
The for would make the condition less obvious.

This way when reading, you know exactly that nothing will go on for the condition. In the for you would have stuff before that would never execute and would just cause confusion.